tools/wptrunner/wptrunner/executors/platformaccessibility/executorwindowsaccessibility.py

```diff
@@ -0,0 +1,149 @@
+# mypy: allow-untyped-defs
+
+from ..platformaccessibility.ia2.constants import *  # noqa: F403
+
+import time
+
+import ctypes
+from ctypes import POINTER, byref
+from ctypes.wintypes import BOOL, HWND, LPARAM
+
+import comtypes.client
+from comtypes import IServiceProvider
+
+CHILDID_SELF = 0
+OBJID_CLIENT = -4
+
+user32 = ctypes.windll.user32  # type: ignore
+oleacc = ctypes.oledll.oleacc  # type: ignore
+oleacc_mod = comtypes.client.GetModule("oleacc.dll")
+IAccessible = oleacc_mod.IAccessible  # noqa: N816
+
+# CoCreateInstance of UIA also initializes IA2
+uia_mod = comtypes.client.GetModule("UIAutomationCore.dll")
+uia_client = comtypes.CoCreateInstance(
+    uia_mod.CUIAutomation._reg_clsid_,
+    interface=uia_mod.IUIAutomation,
+    clsctx=comtypes.CLSCTX_INPROC_SERVER,
+)
+
+
+def accessible_object_from_window(hwnd):
+    p = POINTER(IAccessible)()
+    oleacc.AccessibleObjectFromWindow(
+        hwnd, OBJID_CLIENT, byref(IAccessible._iid_), byref(p)
+    )
+    return p
+
+
+def name_from_hwnd(hwnd):
+    MAX_CHARS = 257
+    buffer = ctypes.create_unicode_buffer(MAX_CHARS)
+    user32.GetWindowTextW(hwnd, buffer, MAX_CHARS)
+    return buffer.value
+
+
+def get_browser_hwnd(product_name):
+    found = []
+
+    @ctypes.WINFUNCTYPE(BOOL, HWND, LPARAM)
+    def check_window_name(hwnd, lParam):  # noqa: N803
+        window_name = name_from_hwnd(hwnd)
+        if product_name not in window_name.lower():
+            # EnumWindows should continue enumerating
+            return True
+        found.append(hwnd)
+        # EnumWindows should stop enumerating (since we found the right window)
+        return False
+
+    user32.EnumWindows(check_window_name, LPARAM(0))
+    if not found:
+        raise LookupError(f"Couldn't find {product_name} HWND")
+    return found[0]
+
+
+def to_ia2(node):
+    service = node.QueryInterface(IServiceProvider)
+    return service.QueryService(IAccessible2._iid_, IAccessible2)  # noqa: F405
+
+
+def find_browser(product_name):
+    hwnd = get_browser_hwnd(product_name)
+    root = accessible_object_from_window(hwnd)
+    return to_ia2(root)
+
+
+def poll_for_tab(url, root):
+    tab = find_tab(url, root)
+    while not tab:
+        time.sleep(0.01)
+        tab = find_tab(url, root)
+    return tab
+
+
+def find_tab(url, root):
+    for i in range(1, root.accChildCount + 1):

```

Comment:
Any reason this isn't using the embeds relation like it does for ATK? Recursively searching the tree isn't ideal. To do this:

1. Change the `to_ia2` function to use IAccessible2_2 instead of IAccessible2.
2. Define the IA2_RELATION_EMBEDS constant, since it's not part of the tlb and thus not available to Python:
    `IA2_RELATION_EMBEDS = "embeds"`
3. Do something like this:
    ```
    targets, count = root.relationTargetsOfType(IA2_RELATION_EMBEDS, 1)
    if count >= 1 and targets[0].accValue(CHILDID_SELF) == url:
        return targets[0]
    ```

Reply:
Thank you for the suggestion! I will make this change, but not until after TPAC when I'm back to the machine set up for windows development.